File: 백준/Gold/18223. 민준이와 마산 그리고 건우/민준이와 마산 그리고 건우.py
from sys import stdin
from heapq import heappush, heappop

input = stdin.readline

INF = float('inf')
V, E, P = map(int, input().split())  # V: 정점의 개수, E: 간선의 개수, P: 건우의 위치
vertex = [[] for _ in range(V + 1)]

for _ in range(E):
    a, b, c = map(int, input().split())  # a <-> b: 가중치 c
    vertex[a].append((b, c))
    vertex[b].append((a, c))


def dijkstra(s, e):
    dist = [INF] * (V + 1)
    dist[s] = 0  # 출발지: 1번 정점 -> 도착지: V번 정점
    pq = [(0, s)]
    while pq:
        min_dist, cur_v = heappop(pq)

        if dist[cur_v] != min_dist:
            continue

        for nxt_v, nxt_dist in vertex[cur_v]:
            new_dist = dist[cur_v] + nxt_dist

            if new_dist < dist[nxt_v]:
                dist[nxt_v] = new_dist
                heappush(pq, (new_dist, nxt_v))

    return dist[e]


# 마산까지 최단거리와 건우를 거쳐서 마산을 가는 최단거리가 같다면 -> 건우를 구한다.
if dijkstra(1, V) == dijkstra(1, P) + dijkstra(P, V):
    print("SAVE HIM")
# 건우를 구하는 최단거리와 전체 최단거리가 다르다면 -> 건우를 구하지 못한다.
else:
    print("GOOD BYE")

File: Java/백준/Gold/18223. 민준이와 마산 그리고 건우/민준이와 마산 그리고 건우.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

class Main {

  static class Node {

    int nxt;
    int dist;

    Node(int nxt, int dist) {
      this.nxt = nxt;
      this.dist = dist;
    }
  }

  private static final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
  private static final StringBuilder sb = new StringBuilder();
  private static StringTokenizer st;

  private static int v, e, p, a, b, c;
  private static final int INF = 987_654_321;
  private static final List<List<Node>> graph = new ArrayList<>();

  public static void main(String[] args) throws Exception {
    init();
    solve();
  }

  private static void init() throws Exception {
    st = new StringTokenizer(br.readLine());
    v = Integer.parseInt(st.nextToken());
    e = Integer.parseInt(st.nextToken());
    p = Integer.parseInt(st.nextToken());

    for (int i = 0; i <= v; i++) {
      graph.add(new ArrayList<>());
    }

    for (int i = 0; i < e; i++) {
      st = new StringTokenizer(br.readLine());
      a = Integer.parseInt(st.nextToken());
      b = Integer.parseInt(st.nextToken());
      c = Integer.parseInt(st.nextToken());
      // a <-> b: 양방향 c
      graph.get(a).add(new Node(b, c));
      graph.get(b).add(new Node(a, c));
    }

    br.close();
  }

  private static void solve() {

    int startToMasan = dijkstra(1, v);
    int startToP = dijkstra(1, p);
    int pToMasan = dijkstra(p, v);

    System.out.print(startToMasan == startToP + pToMasan ? "SAVE HIM" : "GOOD BYE");
  }

  private static int dijkstra(int s, int e) {

    PriorityQueue<int[]> pq = new PriorityQueue<>((o1, o2) -> Integer.compare(o1[1], o2[1]));
    int[] dist = new int[v + 1];
    Arrays.fill(dist, INF);

    dist[s] = 0;
    pq.offer(new int[]{s, 0});

    while (!pq.isEmpty()) {

      int[] curr = pq.poll();
      int node = curr[0];
      int cost = curr[1];
      // 갱신이 이루어지지 않은 경우 -> continue
      if (dist[curr[0]] != cost) {
        continue;
      }

      for (Node nxtNode : graph.get(node)) {
        int newDist = dist[node] + nxtNode.dist;
        if (newDist < dist[nxtNode.nxt]) {
          dist[nxtNode.nxt] = newDist;
          pq.offer(new int[]{nxtNode.nxt, newDist});
        }
      }
    }

    return dist[e];
  }
}